lib/TPP/ConvertPackOptimization.cpp

```diff
@@ -0,0 +1,223 @@
+        }
+      }
+    }
+
+    auto zero = getConstIndex(rewriter, 0);
+    auto one = getConstIndex(rewriter, 1);
+
+    SmallVector<Value> lbs;
+    SmallVector<Value> ubs;
+    SmallVector<Value> steps;
+
+    for (int i = 0; i < numLoops; i++) {
+      lbs.push_back(zero);
+      steps.push_back(one);
+    }
+
+    for (int i = 0; i < numLoops; i++) {
+
+      if (haveMatched[i].haveMatch) {
+        ubs.push_back(getConstIndex(
+            rewriter,
+            shape[i] / packOp.getStaticInnerTiles()[haveMatched[i].index]));
+
+      } else {
+        ubs.push_back(getConstIndex(rewriter, shape[i]));
+      }
+    }
+    Value pos;
+
+    SmallVector<Value> reduc = {
+        packOp.getDest(),
+    };
+
+    auto bodyBuilder = [&](OpBuilder &builder, Location, Value iv,
+                           MutableArrayRef<Value> reduc) {};
+
+    auto loopNest = mlir::scf::buildLoopNest(
+        rewriter, packOp.getLoc(), lbs, ubs, steps, reduc,
+        [&reduc, &pos, bodyBuilder, &packOp,
+         &numLoops](OpBuilder &rewriter, Location loc, ValueRange localIvs,
+                    ValueRange iterArgs) -> scf::ValueVector {
+          reduc.assign(iterArgs.begin(), iterArgs.end());
+
+          SmallVector<OpFoldResult> offsets;
+          SmallVector<MatchResult> haveMatched = innerDimMatchesIndex(packOp);
+          for (int i = 0; i < numLoops; i++) {
+
+            if (haveMatched[i].haveMatch) {
+              Value muliOp = rewriter.create<arith::MulIOp>(
+                  loc, localIvs[i],
+                  getConstIndex(
+                      rewriter,
+                      packOp.getStaticInnerTiles()[haveMatched[i].index]));
+              offsets.push_back(muliOp);
+            } else {
+              offsets.push_back(localIvs[i]);
+            }
+          }
+          SmallVector<OpFoldResult> strides;
+          for (int i = 0; i < numLoops; i++) {
+            strides.push_back(rewriter.getIndexAttr(1));
+          }
+
+          SmallVector<OpFoldResult> sizes;
+
+          for (int i = 0; i < numLoops; i++) {
+            if (haveMatched[i].haveMatch) {
+              sizes.push_back(rewriter.getIndexAttr(
+                  packOp.getStaticInnerTiles()[haveMatched[i].index]));
+            } else {
+              sizes.push_back(rewriter.getIndexAttr(1));
+            }
+          }
+
+          auto tensorExtractType =
+              tensor::ExtractSliceOp::inferCanonicalRankReducedResultType(
+                  packOp.getStaticInnerTiles().size(),
+                  packOp.getSource().getType().cast<RankedTensorType>(),
+                  offsets, sizes, strides);
+
+          auto tensorExtract = rewriter.create<tensor::ExtractSliceOp>(
+              loc, tensorExtractType.cast<RankedTensorType>(),
+              packOp.getSource(), offsets, sizes, strides);
+
+          SmallVector<OpFoldResult> insertSliceOffsets;
+          for (int i = 0; i < numLoops; i++) {
+            int indirection = i;
```

Comment:
what is `indirection`?

Reply:
indirection refers to the index used to index based on outer_dims_perm. Since it is an indirect access through local induction variable, I have named it so.